core_v2/sources/v2_domains.move

```diff
@@ -0,0 +1,1225 @@
+    }
+
+    /// Creates a token for the name.
+    /// NOTE: This function performs no validation checks
+    fun create_token(
+        to_addr: address,
+        domain_name: String,
+        subdomain_name: Option<String>,
+        expiration_time_sec: u64,
+    ) acquires CollectionCapability {
+        let name = v2_token_helper::get_fully_qualified_domain_name(subdomain_name, domain_name);
+        let description = v2_config::tokendata_description();
+        let uri: string::String = v2_config::tokendata_url_prefix();
+        string::append(&mut uri, name);
+
+        let subdomain_ext: Option<SubdomainExt>;
+        let constructor_ref = token::create_named_token(
+            &get_token_signer(),
+            get_collection_name(is_subdomain(subdomain_name)),
+            description,
+            name,
+            option::none(),
+            uri,
+        );
+        let token_signer = object::generate_signer(&constructor_ref);
+        // creating subdomain
+        if (is_subdomain(subdomain_name)) {
+            subdomain_ext = option::some(SubdomainExt {
+                subdomain_name: option::extract(&mut subdomain_name),
+                subdomain_expiration_policy: SUBDOMAIN_POLICY_MANUAL_SET_EXPIRATION,
+            });
+        }
+        // creating domain
+        else {
+            subdomain_ext = option::none<SubdomainExt>();
+        };
+        let record = NameRecord {
+            domain_name,
+            expiration_time_sec,
+            target_address: option::none(),
+            transfer_ref: object::generate_transfer_ref(&constructor_ref),
+            extend_ref: object::generate_extend_ref(&constructor_ref),
+            subdomain_ext,
+        };
+        move_to(&token_signer, record);
+        let record_obj = object::object_from_constructor_ref<NameRecord>(&constructor_ref);
+        object::transfer(&get_token_signer(), record_obj, to_addr);
+    }
+
+    // === REGISTER NAME ===
+
+    /// A wrapper around `register_name` as an entry function.
+    /// Option<String> is not currently serializable, so we have these convenience methods
+    public fun register_domain(
+        router_signer: &signer,
+        sign: &signer,
+        domain_name: String,
+        registration_duration_secs: u64,
+    ) acquires CollectionCapability, NameRecord, RegisterNameEvents, ReverseRecord, SetReverseLookupEvents {
+        assert!(address_of(router_signer) == @router_signer, error::permission_denied(ENOT_ROUTER));
+        assert!(v2_config::unrestricted_mint_enabled(), error::permission_denied(EVALID_SIGNATURE_REQUIRED));
+
+        validate_registration_duration(registration_duration_secs);
+
+        let subdomain_name = option::none<String>();
+
+        assert!(is_name_registerable(domain_name, subdomain_name), error::invalid_state(ENAME_NOT_AVAILABLE));
+
+        let length = validate_name_string(domain_name);
+
+        let price = v2_price_model::price_for_domain(length, registration_duration_secs);
```

Comment:
considering what we discussed, I think this needs to be a flat fee and that we likely bump `registration_duration_secs` to the nearest next chunk of time`

I would be unhappy to pay full price and not get the full re-registration time allotted to me.